modules/mapper-extras/src/main/java/org/elasticsearch/index/mapper/extras/MatchOnlyTextFieldMapper.java

```diff
@@ -204,9 +204,9 @@ private IOFunction<LeafReaderContext, CheckedIntFunction<List<Object>, IOExcepti
                     };
                 };
             }
-            ValueFetcher valueFetcher = valueFetcher(searchExecutionContext, null);
-            SourceProvider sourceProvider = searchExecutionContext.lookup();
             return context -> {
+                ValueFetcher valueFetcher = valueFetcher(searchExecutionContext, null);
+                SourceProvider sourceProvider = searchExecutionContext.lookup();
```

Comment:
For my understanding: this should also take care of the concurrency issue from #100074 since the the `searchExecutionContext.lookup()` call is now done per leaf? Is that the correct reading of this? 
In any case, can you also see if the StoredFieldLoader from L198 should also be moved into the context-lambda in the case where we have synthetic source enabled? I'm only guessing here so maybe it's not needed, but I also guess we don't have tests covering that code path.

Reply:
Yes, that's correct.  We don't need to do the same with StoredFieldLoader as that is meant to be global and then create separate LeafStoredFieldLoaders per-segment.  The issue here is that we have a top-level SourceProvider which is caching segment information.

Reply:
I'll add a test specifically for the synthetic source path as well.

Reply:
It's quite tricky to read the difference here, especiall as valueFetcher does not take any different argument than before :)